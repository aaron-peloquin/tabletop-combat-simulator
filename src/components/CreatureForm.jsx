import React from "react"
import {Button,Grid,TextField} from "@material-ui/core/"
import { withStyles } from "@material-ui/core/styles"

import standardizeCreatureData from "./../helpers/standardizeCreatureData"

let data, originalData, _cbSubmit, _cbUpdate

const styles = (theme) => {
  return {
    formField: {
      width:"100%",
      marginTop: theme.spacing.unit * 2,
      paddingLeft: theme.spacing.unit,
      paddingRight: theme.spacing.unit,
    }
  }
}

/** Handle updating data */
const updateData = (e, key) => {
  const curValue = e.target.value||""
  data[key] = curValue

  if(typeof _cbUpdate=="function") {
    _cbUpdate(key,curValue)
  }
}

/** Handle the form submissions callback */
const submitData = (e) => {
  e.preventDefault()
  if(data.name.length>0 && typeof _cbSubmit=="function") {
    _cbSubmit(data)
  }
  data = originalData
}

/** Builds <Grid> input fields for the <CreatureForm /> element */
let GridField = (props) => {
  const {classes, dataKey, field, label, size} = props
  let wrappedField, DataField

  let attributes = {
    onChange: (e)=>{ updateData(e, dataKey) },
    className: classes.formField,
    defaultValue: data[dataKey],
    label: label
  }
  if(field=="multiline") {
    DataField = <TextField multiline rows="3" {...attributes} />
  }
  else {
    DataField = <TextField {...attributes} />
  }

  switch(size) {
  default:
  case "small":
    wrappedField = <Grid item xs={12} sm={6} md={3}>{DataField}</Grid>
    break
  case "medium":
    wrappedField = <Grid item xs={12} sm={6}>{DataField}</Grid>
    break
  case "large":
    wrappedField = <Grid item xs={12}>{DataField}</Grid>
    break
  }

  return wrappedField
}
GridField = withStyles(styles)(GridField)

const CreatureForm = (props) => {
  const {creature={}} = props
  /** If this object does not has a hash, set it to null so we generate one */
  if(!creature.hasOwnProperty("hash")) {
    creature.hash = null
  }
  data = standardizeCreatureData(props.creature)
  originalData = Object.assign({}, data)
  _cbSubmit = props.onSubmit||function(){}
  _cbUpdate = props.onUpdate||function(){}
  return <form noValidate autoComplete="off" onSubmit={submitData}>
    <fieldset>
      <legend>Creature</legend>
      <Grid container>
        <GridField size="large" dataKey="name" label="Name" />
        <GridField size="large" field="multiline" dataKey="description" label="Description" />

        <GridField dataKey="cr" label="Challenge Rating" />
        <GridField dataKey="hp" label="Hit Points" />
        <GridField dataKey="armor" label="Hit Points" />
        <GridField dataKey="defaultInitiative" label="Default Initiative" />

        <GridField size="medium" dataKey="hitDiceEquation" label="Hit Dice Equation" placeholder="eg. 1d20+4" />
        <GridField size="medium" dataKey="damageDiceEquation" label="Damage Dice Equation" placeholder="eg. 1d6+2d4+2" />
      </Grid>
      <Button type="submit" onClick={submitData}>Submit</Button>
    </fieldset>
  </form>
}

export default CreatureForm