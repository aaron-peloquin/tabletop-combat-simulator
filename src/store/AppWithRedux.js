import React from 'react'
import {initializeStore} from './store'

const isServer = typeof window === 'undefined'
const reduxStoreSymbol = Symbol("reduxStore")

function getOrCreateStore (initialState) {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    return initializeStore(initialState)
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[reduxStoreSymbol]) {
    window[reduxStoreSymbol] = initializeStore(initialState)
  }
  return window[reduxStoreSymbol]
}

export default (App) => {
  return class AppWithRedux extends React.Component {
    static async getInitialProps (appContext) {
      // Get or Create the store with `undefined` as initialState
      // This allows you to set a custom default initialState
      const reduxStore = getOrCreateStore()

      // Provide the store to getInitialProps of pages
      appContext.ctx.reduxStore = reduxStore

      let appProps = {}
      if (typeof App.getInitialProps === 'function') {
        appProps = await App.getInitialProps(appContext)
      }

      return {
        ...appProps,
        initialReduxState: reduxStore.getState()
      }
    }

    constructor (props) {
      super(props)
      this.reduxStore = getOrCreateStore(props.initialReduxState)
    }

    render () {
      return <App {...this.props} reduxStore={this.reduxStore} />
    }
  }
}
