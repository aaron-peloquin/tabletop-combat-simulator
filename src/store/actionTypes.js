/**
 * This is an object of Redux Actions. It mostly exists to
 *  ensure we do not create duplicate actions, but it can
 *  also be used as a list of our existing actions
*/
const actions = {
  CopyCreature: 'CopyCreature',
  DeleteAllCreatures: 'DeleteAllCreatures',
  DeleteCreature: 'DeleteCreature',
  SaveCreature: 'SaveCreature',
  SetEditCreature: 'SetEditCreature',
  ToggleSidebar: 'ToggleSidebar',
  RunSimulation: 'RunSimulation',
  UpdateEditCreature: 'UpdateEditCreature',
}

export default actions
