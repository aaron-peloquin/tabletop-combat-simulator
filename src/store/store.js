import {createStore, applyMiddleware, combineReducers} from "redux"
import {composeWithDevTools} from "redux-devtools-extension"
import thunkMiddleware from "redux-thunk"

/** Reducers */
import combat from "./reducer/combat"
import creatures from "./reducer/creatures"
import sideBar from "./reducer/sideBar"
import editing from "./reducer/editing"

/** Reducers */
const reducers = combineReducers({
  combat,
  creatures,
  editing,
  sideBar,
})

/**
 * Initializes the store
 * @param {obj} initialState the state our app starts with
 * @return {obj} the redux store
 */
export function initializeStore(initialState = initialState) {
  return createStore(reducers, initialState, composeWithDevTools(applyMiddleware(thunkMiddleware)))
}
