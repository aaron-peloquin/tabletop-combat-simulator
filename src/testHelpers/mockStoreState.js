export default {
  "creatures": [
    {
      "name": "Goplin",
      "cr": "1/4",
      "hash": "hivn3p0k",
      "description": "A stout monster",
      "defaultInitiative": 10,
      "armor": 11,
      "hp": 5,
      "hitDiceEquation": "1d20+1",
      "damageDiceEquation": "1d4"
    },
    {
      "name": "Elven Ranger",
      "cr": "1/2",
      "hash": "h3dxf5wl",
      "description": "A swift hunter",
      "defaultInitiative": 16,
      "armor": 13,
      "hp": 10,
      "hitDiceEquation": "1d20+8",
      "damageDiceEquation": "1d6+3"
    },
    {
      "name": "Knight",
      "cr": "1",
      "hash": "hee9p286",
      "description": "An honorable soldier",
      "defaultInitiative": 11,
      "armor": 18,
      "hp": 25,
      "hitDiceEquation": "1d20+4",
      "damageDiceEquation": "1d12+4"
    },
    {
      "name": "Lizardfolk",
      "cr": "3",
      "hash": "howx26rt",
      "description": "An ancient sneaky monster",
      "defaultInitiative": 14,
      "armor": 14,
      "hp": 15,
      "hitDiceEquation": "1d20+5",
      "damageDiceEquation": "1d10+2"
    }
  ],
  "sideBar": false,
}