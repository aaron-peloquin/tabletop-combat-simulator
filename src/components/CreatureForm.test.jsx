import { Provider } from "react-redux"
import { createShallow } from "@material-ui/core/test-utils"
import toJson from "enzyme-to-json"
import { initializeStore } from "../store/store"

import CreatureForm from "./CreatureForm"
import mockStoreState from "../testHelpers/mockStoreState"

describe("<CreatureForm />", ()=>{
  const RenderShallowUntilComponent = createShallow({"untilSelector":"CreatureForm"})
  let store, props, Component
  beforeEach(()=>{
    store = initializeStore(mockStoreState)
    props = {
      onSubmit: jest.fn(()=>{}),
      onChange: jest.fn(()=>{}),
      router: { push: jest.fn(() => {}) },
      EditingCreature: mockStoreState.creatures[0],
    }
    Component = RenderShallowUntilComponent(<Provider store={store}><CreatureForm {...props} /></Provider>)
  })

  it("loads", () => {
    expect(typeof Component).toBe("object")
  })

  it("has a <Grid> container", ()=>{
    expect(Component.find("WithStyles(Grid)[container=true]").length).toBeGreaterThanOrEqual(1)
  })

  it("fires submit callback with primary button", ()=>{
    Component.find("WithStyles(Button)[color=\"primary\"]").simulate("click", { preventDefault:()=>{} })
    expect(props.onSubmit).toHaveBeenCalledTimes(1)
  })

  it("fires submit callback with secondary button", ()=>{
    Component.find("WithStyles(Button)[color=\"secondary\"]").simulate("click", { preventDefault:()=>{} })
    expect(props.onSubmit).toHaveBeenCalledTimes(1)
  })

  it("fires change callback", ()=>{
    Component.find("Connect(WithStyles(CreatureFormGridField))[dataKey=\"name\"]").simulate("change")
    expect(props.onChange).toHaveBeenCalledTimes(1)
  })

  it("snapshots", () => {
    expect(toJson(Component)).toMatchSnapshot()
  })
})