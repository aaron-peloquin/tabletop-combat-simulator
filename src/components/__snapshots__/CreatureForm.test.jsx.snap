// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreatureForm /> snapshots 1`] = `
<form
  autoComplete="off"
  onSubmit={[Function]}
>
  <button
    className="CreatureForm-hidden-1"
    type="submit"
  />
  <fieldset>
    <legend>
      Player or Creature
    </legend>
    <WithStyles(Grid)
      container={true}
    >
      <Connect(WithStyles(CreatureFormGridField))
        dataKey="name"
        label="Name (Required)"
        onChange={[Function]}
        size="large"
      />
      <Connect(WithStyles(CreatureFormGridField))
        dataKey="hp"
        label="Starting Hit Points"
        onChange={[Function]}
      />
      <Connect(WithStyles(CreatureFormGridField))
        dataKey="armor"
        label="Armor"
        onChange={[Function]}
      />
      <Connect(WithStyles(CreatureFormGridField))
        dataKey="initiative"
        label="Initiative Bonus"
        onChange={[Function]}
      />
      <Connect(WithStyles(CreatureFormGridField))
        dataKey="hitDiceEquation"
        label="Hit Dice Equation"
        onChange={[Function]}
        placeholder="eg. 1d20+4"
        size="medium"
      />
      <Connect(WithStyles(CreatureFormGridField))
        dataKey="damageDiceEquation"
        label="Damage Dice Equation"
        onChange={[Function]}
        placeholder="eg. 1d6+2d4+2"
        size="medium"
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      container={true}
      justify="center"
    >
      <WithStyles(Grid)
        item={true}
        sm={6}
      >
        <WithStyles(Button)
          className="CreatureForm-submitButtons-2"
          color="primary"
          data-cy="aTeamSubmit"
          onClick={[Function]}
          type="submit"
          variant="contained"
        >
          Save
           to Team A
        </WithStyles(Button)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        sm={6}
      >
        <WithStyles(Button)
          className="CreatureForm-submitButtons-2"
          color="secondary"
          data-cy="bTeamSubmit"
          onClick={[Function]}
          type="submit"
          variant="contained"
        >
          Save
           to Team B
        </WithStyles(Button)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </fieldset>
</form>
`;
